nixos/modules/services/networking/lokinet.nix

```diff
@@ -0,0 +1,219 @@
+{ config, lib, pkgs, ... }:
+
+let
+  cfg = config.services.lokinet;
+
+  snappFile = pkgs.writeTextFile {
+    name = "snapp-example.ini";
+
+    text = ''
+      [example-snapp]
+      keyfile=${cfg.dataDir}/.lokinet/example-snap-keyfile.private
+      ifaddr=${cfg.snapp.ifAddr}
+      ifname=${cfg.snapp.ifName}
+    '' + cfg.snapp.extraConfig;
+  };
+
+  configFile = pkgs.writeTextFile {
+    name = "lokinet.ini";
+
+    text = ''
+      [router]
+      threads=${toString cfg.threads}
+      contact-file=${cfg.dataDir}/.lokinet/self.signed
+      transport-privkey=${cfg.dataDir}/.lokinet/transport.private
+      ident-privkey=${cfg.dataDir}/.lokinet/identity.private
+      encryption-privkey=${cfg.dataDir}/.lokinet/encryption.private
+    '' + lib.optionalString (cfg.nickname != null) ''
+      nickname=${cfg.nickname}
+    '' + ''
+      min-routers=${toString cfg.minRouters}
+      max-routers=${toString cfg.maxRouters}
+
+      [logging]
+      level=info
+      type=syslog
+
+      [metrics]
+      json-metrics-path=${cfg.dataDir}/.lokinet/metrics.json
+
+      [api]
+      enabled=false
+      bind=127.0.0.1:1190
+
+      [system]
+      user=lokinet
+      group=lokinet
+      pidfile=${cfg.dataDir}/.lokinet/lokinet.pid
+
+      [dns]
+      upstream=${cfg.resolverUpstream}
+      bind=${cfg.resolverBind}
+
+      [netdb]
+      dir=${cfg.dataDir}/.lokinet/netdb
+
+      [bootstrap]
+      add-node=${cfg.dataDir}/.lokinet/bootstrap.signed
+
+      [services]
+    '' + lib.optionalString (cfg.snapp != null) ''
+      example-snapp=${cfg.dataDir}/.lokinet/snapp-example.ini
+    '' + ''
+      [network]
+      profiles=${cfg.dataDir}/.lokinet/profiles.dat
+    '' + lib.optionalString (cfg.strictConnect) ''
+      strict-connect=pubkey
+    '' + lib.optionalString (cfg.exitNode) ''
+      exit-node=pubkey
+    '' + ''
+      enabled=true
+    '' + cfg.extraConfig;
+  };
+in with lib; {
+  options.services.lokinet = {
+    enable = mkEnableOption "lokinet";
+
+    dataDir = mkOption {
```

Comment:
I've been told to do `baseDir`, `dataDir`, and now make the directory with systemd with several PRs concerning services. Which one is it?
Also normally this service is run as the user with libcap permissions to alter network features, so having it in any user-defined directory should be absolutely fine.

Reply:
@Chiiruno I can appreciate your frustration on receiving conflicting direction on the matter. There isn't an official consensus so you run into people pushing their opinion on you.

Specifically in this case I mention this to be problematic because `tmpfiles.d` will throw an error with the example you gave and the service won't be able to start. `tmpfiles.d` won't allow you to create a directory owned by `lokinet` inside a folder owned by `randomuser` as this is not considered safe.

Reply:
Okay, that makes sense.
So I just get rid of `dataDir` and use `StateDirectory` in `ServiceConfig`, IIRC?
Is there no way to still allow a user-defined directory with that?

Reply:
This is a bit of a problem with `systemd` uptake in NixOS I've realized. Fundamentally `systemd` differentiates between `system` and `user` services. Some of our modules were written to not respect this difference, and the prevalence of `home-manager` and it's great modules has made the problem go relatively unnoticed. For example, our `mpd` module is broken in that people run it as a regular user account (it doesn't even work otherwise) but this is a problem because the module runs it as a `systemd` system service instead of a `user` service. Wow I guess I went on a bit of a rant there... :smile:

To tie this back into your PR: if you want the data directory inside your home directory are you sure this is a `systemd` `system` service and not a `user` service?

Reply:
Originally, this program is meant to be ran as the user with libcap permissions, so I wanted to at least give people the choice of where they want to put the directory.
AFAIK, it'll work either way, but I'm all about giving choices.

Reply:
If the program can run as a dedicated user then a `systemd` `system` service is the way to go, in my opinion. If not then you might want to look into `systemd` `user` service with something like `security.wrappers`... or ask someone who knows more than I do on this topic (hint: I don't know that much). :man_shrugging: